### ISSUE TITLE
fail install env

### ISSUE THREAD

**Issue by HernandoR** [astral-sh/ruff-pre-commit#125]
please share any idea for the following error log

pre-commit-config
```
repos:
- repo: https://github.com/astral-sh/ruff-pre-commit
  # Ruff version.
  rev: v0.9.3
  hooks:
  # # Run the linter.
  - id: ruff
    types_or: [ python, pyi ]
    args: [ '--fix' ]
  # Run the formatter.
  - id: ruff-format
    types_or: [ python, pyi ]
```
logs:
```
++ pre-commit run --all-files
[INFO] Installing environment for https://github.com/astral-sh/ruff-pre-commit.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('/root/.cache/pre-commit/repojf50w8xb/py_env-python3.9/bin/python', '-mpip', 'install', '.')
return code: 1
expected return code: 0
stdout:
    Processing /root/.cache/pre-commit/repojf50w8xb
      Installing build dependencies: started
      Installing build dependencies: finished with status 'error'
    
stderr:
      ERROR: Command errored out with exit status 2:
       command: /root/.cache/pre-commit/repojf50w8xb/py_env-python3.9/bin/python /tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-npqatk8q/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=40.8.0' wheel
           cwd: None
      Complete output (80 lines):
      Collecting setuptools>=40.8.0
        Downloading setuptools-80.4.0-py3-none-any.whl (1.2 MB)
      ERROR: Exception:
      Traceback (most recent call last):
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_vendor/urllib3/response.py", line 438, in _error_catcher
          yield
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_vendor/urllib3/response.py", line 519, in read
          data = self._fp.read(amt) if not fp_closed else b""
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_vendor/cachecontrol/filewrapper.py", line 62, in read
          data = self.__fp.read(amt)
        File "/usr/local/lib/python3.9/http/client.py", line 463, in read
          n = self.readinto(b)
        File "/usr/local/lib/python3.9/http/client.py", line 507, in readinto
          n = self.fp.readinto(b)
        File "/usr/local/lib/python3.9/socket.py", line 704, in readinto
          return self._sock.recv_into(b)
        File "/usr/local/lib/python3.9/ssl.py", line 1241, in recv_into
          return self.read(nbytes, buffer)
        File "/usr/local/lib/python3.9/ssl.py", line 1099, in read
          return self._sslobj.read(len, buffer)
      socket.timeout: The read operation timed out
      
      During handling of the above exception, another exception occurred:
      
      Traceback (most recent call last):
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/cli/base_command.py", line 164, in exc_logging_wrapper
          status = run_func(*args)
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/cli/req_command.py", line 205, in wrapper
          return func(self, options, args)
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/commands/install.py", line 338, in run
          requirement_set = resolver.resolve(
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/resolution/resolvelib/resolver.py", line 92, in resolve
          result = self._result = resolver.resolve(
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_vendor/resolvelib/resolvers.py", line 482, in resolve
          state = resolution.resolve(requirements, max_rounds=max_rounds)
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_vendor/resolvelib/resolvers.py", line 349, in resolve
          self._add_to_criteria(self.state.criteria, r, parent=None)
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_vendor/resolvelib/resolvers.py", line 173, in _add_to_criteria
          if not criterion.candidates:
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_vendor/resolvelib/structs.py", line 151, in __bool__
          return bool(self._sequence)
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/resolution/resolvelib/found_candidates.py", line 155, in __bool__
          return any(self)
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/resolution/resolvelib/found_candidates.py", line 143, in <genexpr>
          return (c for c in iterator if id(c) not in self._incompatible_ids)
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/resolution/resolvelib/found_candidates.py", line 47, in _iter_built
          candidate = func()
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/resolution/resolvelib/factory.py", line 201, in _make_candidate_from_link
          self._link_candidate_cache[link] = LinkCandidate(
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/resolution/resolvelib/candidates.py", line 281, in __init__
          super().__init__(
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/resolution/resolvelib/candidates.py", line 156, in __init__
          self.dist = self._prepare()
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/resolution/resolvelib/candidates.py", line 225, in _prepare
          dist = self._prepare_distribution()
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/resolution/resolvelib/candidates.py", line 292, in _prepare_distribution
          return preparer.prepare_linked_requirement(self._ireq, parallel_builds=True)
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/operations/prepare.py", line 482, in prepare_linked_requirement
          return self._prepare_linked_requirement(req, parallel_builds)
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/operations/prepare.py", line 527, in _prepare_linked_requirement
          local_file = unpack_url(
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/operations/prepare.py", line 213, in unpack_url
          file = get_http_url(
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/operations/prepare.py", line 94, in get_http_url
          from_path, content_type = download(link, temp_dir.path)
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/network/download.py", line 145, in __call__
          for chunk in chunks:
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/cli/progress_bars.py", line 144, in iter
          for x in it:
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_internal/network/utils.py", line 63, in response_chunks
          for chunk in response.raw.stream(
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_vendor/urllib3/response.py", line 576, in stream
          data = self.read(amt=amt, decode_content=decode_content)
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_vendor/urllib3/response.py", line 541, in read
          raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
        File "/usr/local/lib/python3.9/contextlib.py", line 137, in __exit__
          self.gen.throw(typ, value, traceback)
        File "/tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip/_vendor/urllib3/response.py", line [443](https://git.n.xiaomi.com/mi-autolabel/4dod/-/jobs/33634017#L443), in _error_catcher
          raise ReadTimeoutError(self._pool, None, "Read timed out.")
      pip._vendor.urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out.
      ----------------------------------------
    WARNING: Discarding file:///root/.cache/pre-commit/repojf50w8xb. Command errored out with exit status 2: /root/.cache/pre-commit/repojf50w8xb/py_env-python3.9/bin/python /tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-npqatk8q/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=40.8.0' wheel Check the logs for full command output.
    ERROR: Command errored out with exit status 2: /root/.cache/pre-commit/repojf50w8xb/py_env-python3.9/bin/python /tmp/pip-standalone-pip-9tt7keuj/__env_pip__.zip/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-npqatk8q/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=40.8.0' wheel Check the logs for full command output.
```

**Comment by MichaReiser** [comment]
This looks like a temporary network problem. Can you try to re-run the command

**Comment by HernandoR** [comment]
it consists, that runner is in a limited network (though was working fine for the past months). But I could not make it work in another machine with unlimited network. error log is as follows
```
[INFO] Installing environment for https://github.com/astral-sh/ruff-pre-commit.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('/home/lz/.cache/pre-commit/repo_3rqksaj/py_env-python3/bin/python', '-mpip', 'install', '.')
return code: 1
stdout: (none)
stderr:
    Traceback (most recent call last):
      File "<frozen runpy>", line 198, in _run_module_as_main
      File "<frozen runpy>", line 88, in _run_code
      File "/home/lz/.cache/pre-commit/repo_3rqksaj/py_env-python3/lib/python3.12/site-packages/pip/__main__.py", line 24, in <module>
        sys.exit(_main())
                 ^^^^^^^
      File "/home/lz/.cache/pre-commit/repo_3rqksaj/py_env-python3/lib/python3.12/site-packages/pip/_internal/cli/main.py", line 77, in main
        command = create_command(cmd_name, isolated=("--isolated" in cmd_args))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/lz/.cache/pre-commit/repo_3rqksaj/py_env-python3/lib/python3.12/site-packages/pip/_internal/commands/__init__.py", line 114, in create_command
        module = importlib.import_module(module_path)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/python3.12/importlib/__init__.py", line 90, in import_module
        return _bootstrap._gcd_import(name[level:], package, level)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
      File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
      File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
      File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
      File "<frozen importlib._bootstrap_external>", line 995, in exec_module
      File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
      File "/home/lz/.cache/pre-commit/repo_3rqksaj/py_env-python3/lib/python3.12/site-packages/pip/_internal/commands/install.py", line 15, in <module>
        from pip._internal.cli.req_command import (
      File "/home/lz/.cache/pre-commit/repo_3rqksaj/py_env-python3/lib/python3.12/site-packages/pip/_internal/cli/req_command.py", line 21, in <module>
        from pip._internal.index.package_finder import PackageFinder
      File "/home/lz/.cache/pre-commit/repo_3rqksaj/py_env-python3/lib/python3.12/site-packages/pip/_internal/index/package_finder.py", line 30, in <module>
        from pip._internal.req import InstallRequirement
      File "/home/lz/.cache/pre-commit/repo_3rqksaj/py_env-python3/lib/python3.12/site-packages/pip/_internal/req/__init__.py", line 8, in <module>
        from .req_install import InstallRequirement
      File "/home/lz/.cache/pre-commit/repo_3rqksaj/py_env-python3/lib/python3.12/site-packages/pip/_internal/req/req_install.py", line 32, in <module>
        from pip._internal.operations.build.metadata import generate_metadata
    ModuleNotFoundError: No module named 'pip._internal.operations.build.metadata'
Check the log at /home/lz/.cache/pre-commit/pre-commit.log
```

Is it possible to set the index url? Such that we could solve part of the problem

**Comment by MichaReiser** [comment]
>     ModuleNotFoundError: No module named 'pip._internal.operations.build.metadata'

This is a different error from what you see on the "limited network" host. 

I don't there's a way to specify a custom index. But I might be wrong, I'm not very familiar with pre-commit

**Comment by HernandoR** [comment]
I believe it was a temporary network failure.

An example of custom index is shown as follows:
```
repos:
- repo: https://github.com/astral-sh/ruff-pre-commit
  # Ruff version.
  rev: v0.11.9
  hooks:
  # # Run the linter.
  - id: ruff
    additional_dependencies:
      - "--index-url=https://mirrors.cernet.edu.cn/pypi/web/simple"
    types_or: [ python, pyi ]
    args: [ '--fix' ]
  # Run the formatter.
  - id: ruff-format
    additional_dependencies:
      - "--index-url=https://mirrors.cernet.edu.cn/pypi/web/simple"
    types_or: [ python, pyi ]
```
